### PR TITLE
[`flake8`]: Implement `flake8-variables-names`

### PR DESCRIPTION
**Pull request by latonis**
Implementing [flake8-variables-names](https://github.com/best-doctor/flake8-variables-names)

Referenced and requested in #3463

### CODE REVIEW

**Comment by latonis** [comment]
Adding the `--use-varnames-strict-mode` CLI setting currently. Is there a convention for how the CLI options should be defined or can this be ported over directly?

File: crates/ruff/src/registry.rs [hunk 1]
```diff
@@ -754,6 +757,9 @@ pub enum Linter {
     /// [flake8-use-pathlib](https://pypi.org/project/flake8-use-pathlib/)
     #[prefix = "PTH"]
     Flake8UsePathlib,
+    /// [flake8-variables-names](https://pypi.org/project/flake8-variables-names/)
+    #[prefix = "VN"]
```
Comment (charliermarsh) [inline crates/ruff/src/registry.rs:761]: Nit: I think these should be `VNE` to match the upstream plugin -- or was this an intentional deviation?

File: crates/ruff/src/rules/flake8_variables_names/settings.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    rename = "Flake8VariablesNamesOptions"
+)]
+pub struct Options {
+    #[option(
+        default = "false",
+        value_type = "bool",
+        example = "use-varnames-strict-mode = true"
+    )]
+    /// Whether to use strict settings for the allow/block lists
+    pub use_varnames_strict_mode: Option<bool>,
```
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_variables_names/settings.rs:22]: We can probably just shorten this to `strict`, since unlike in Flake8, this is scoped under the plugin settings, which makes it clear that it applies to `flake8-variable-names`.

File: crates/ruff/src/registry.rs (hunk 1 above)
Comment (latonis) [reply crates/ruff/src/registry.rs:761]: I originally had that but changed it due to the few above not having E in their prefix, but makes sense to keep it to the original 🙂

File: crates/ruff/src/rules/flake8_variables_names/settings.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    rename_all = "kebab-case",
+    rename = "Flake8VariablesNamesOptions"
+)]
+pub struct Options {
+    #[option(
+        default = "false",
+        value_type = "bool",
+        example = "use-varnames-strict-mode = true"
+    )]
+    /// Whether to use strict settings for the allow/block lists
```
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_variables_names/settings.rs:21]: Can we add some notes on the effect that strict `True` vs. `False` has on the checks?

**Review by charliermarsh — COMMENTED** [review]
This is really good work. My only hesitation here is that this strikes me as a very strict set of rules (e.g., I don't think I'd enable these on my own projects). But, we're yet to define clear criteria for rule and plugin inclusion, and our current approach has generally been not to avoid injecting too much subjectivity into the plugins we choose to incorporate.

For reference, the factors we'd laid out in the past were: (1) some evidence of usefulness / popularity, (2) non-overlapping behavior with existing plugins, (3) reasonably accurate (few false positives and negatives), and (4) complexity doesn't outweigh its usefulness. Though we don't have formal criteria yet. So, I'm open to incorporating. The main "cost" is that we'll likely surface a bunch of violations to those with `--select ALL` enabled, and they'll be forced to disable or fix.

@MichaReiser is thinking about some of these questions, so \cc'ing him :)

**Comment by latonis** [comment]
> This is really good work. My only hesitation here is that this strikes me as a very strict set of rules (e.g., I don't think I'd enable these on my own projects). But, we're yet to define clear criteria for rule and plugin inclusion, and our current approach has generally been not to avoid injecting too much subjectivity into the plugins we choose to incorporate.
> 
> For reference, the factors we'd laid out in the past were: (1) some evidence of usefulness / popularity, (2) non-overlapping behavior with existing plugins, (3) reasonably accurate (few false positives and negatives), and (4) complexity doesn't outweigh its usefulness. Though we don't have formal criteria yet. So, I'm open to incorporating. The main "cost" is that we'll likely surface a bunch of violations to those with `--select ALL` enabled, and they'll be forced to disable or fix.
> 
> @MichaReiser is thinking about some of these questions, so \cc'ing him :)

Hey Charlie, absolutely understand the concern for how strict it is, I had the same concern after running it and seeing the ecosystem results. No worries at all if this one is too strict and we don't want to include it at this time (or ever). I'll defer to you if you'd like to keep this open or close it, but I'll update the following requests above and implement the changes requested just in-case it is merged in at some point. :)

File: crates/ruff/src/rules/flake8_variables_names/settings.rs (hunk 2 above)
Comment (latonis) [reply crates/ruff/src/rules/flake8_variables_names/settings.rs:22]: Makes sense to me!

File: crates/ruff/src/rules/flake8_variables_names/settings.rs (hunk 3 above)
Comment (latonis) [reply crates/ruff/src/rules/flake8_variables_names/settings.rs]: definitely; it has been updated

File: crates/ruff/src/rules/flake8_variables_names/rules/single_letter_variable_name.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        let SingleLetterVariableName(name) = self;
+        format!("Single letter variable name: `{name}`")
+    }
+}
+
+fn is_single_letter_variable(name: &str, strict_mode: bool) -> bool {
+    const ALLOWLIST: [&str; 3] = ["i", "_", "T"];
+    const ALLOWLIST_STRICT: [&str; 2] = ["_", "T"];
+    if name.len() == 1
+        && (!ALLOWLIST.contains(&name) || (strict_mode && !ALLOWLIST_STRICT.contains(&name)))
```
Comment (MichaReiser) [inline crates/ruff/src/rules/flake8_variables_names/rules/single_letter_variable_name.rs:35]: Nit: you can write this with a match
```suggestion
        && (!matches!(name, "i", "_", "T") || (strict_mode && !matches!(name, "_", "T)))
```

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff/src/rules/flake8_variables_names/rules/single_letter_variable_name.rs (hunk 4 above)
Comment (latonis) [reply crates/ruff/src/rules/flake8_variables_names/rules/single_letter_variable_name.rs:35]: > Nit: you can write this with a match

```
(!matches!(name, ["i", "_", "T"]) || (strict_mode && !matches!(name, "_", "T")))   |                                                                                      ^^^ no rules expected this token in macro call   |   = note: while trying to match sequence end
```

Currently getting this when implementing as a match statement

File: crates/ruff/src/rules/flake8_variables_names/rules/non_descript_variable_name.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    const BLOCKLIST_STRICT: [&str; 11] = [
+        "data", "result", "results", "item", "items", "value", "values", "content", "obj", "info",
+        "handler",
+    ];
+
+    if BLOCKLIST.contains(&name.to_lowercase().as_str())
+        || (strict_mode && BLOCKLIST_STRICT.contains(&name.to_lowercase().as_str()))
+    {
+        return true;
+    }
```
[the remark this answers is not in the document]
Comment (latonis) [reply crates/ruff/src/rules/flake8_variables_names/rules/non_descript_variable_name.rs:49]: Implemented :)

File: crates/ruff/src/rules/flake8_variables_names/rules/single_letter_variable_name.rs (hunk 4 above)
Comment (latonis) [reply crates/ruff/src/rules/flake8_variables_names/rules/single_letter_variable_name.rs:35]: Ah I see what I did incorrectly

**Comment by latonis** [comment]
pre-commit failed due to my own rules
image

**Comment by charliermarsh** [comment]
Sorry for letting this sit @latonis, and for not giving clearer feedback prior to working on the issue. There is actually a chance we re-open this in the future, since it is a popular plugin... But I will take ownership of rebasing etc. if that time comes.

**Comment by latonis** [comment]
No worries at all @charliermarsh, I closed per our previous conversations about not being sure about too strict of rules being included when running all of them and that sort. 

Happy to re-do the work if the time comes, I just figured with the amount of improvements ruff has had internally it would probably be easier to re-implement the logic rather than try to re-base it 🙂
